kong/tools/queue.lua

```diff
@@ -0,0 +1,400 @@
+--       max_retry_delay = 60,   -- maximum delay between send attempts, caps exponential retry
+--     }
+--
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Some value")
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Another value")
+--
+-- Given the example above,
+--
+-- * If the two `enqueue()` invocations are done within `max_coalescing_delay` seconds, they will be passed to the
+--   handler function together in one batch.  The maximum number of entries in one batch is defined
+--   by the `max_batch_size` parameter.
+-- * The `max_entries` parameter defines how many entries can be waiting on the queue for transmission.  If
+--   it is exceeded, the oldest entries on the queue will be discarded when new entries are queued.  Error
+--   messages describing how many entries were lost will be logged.
+-- * The `max_bytes` parameter, if set, indicates that no more than that number of bytes can be
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
+-- * When the `handler` function does not return a true value for a batch, it is retried for up to
+--   `max_retry_time` seconds before the batch is deleted and an error is logged.  Retries are organized
+--   by the queue library using an exponential back-off algorithm with the maximum time between retries
+--   set to `max_retry_delay` seconds.
+
+local workspaces = require "kong.workspaces"
+local semaphore = require "ngx.semaphore"
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
+end
+
+
+local Queue = {
+  CAPACITY_WARNING_THRESHOLD = 0.8, -- Threshold to warn that the queue max_entries limit is reached
+}
+
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+local function make_queue_key(name)
+  return string.format("%s.%s", workspaces.get_workspace_id(), name)
+end
+
+
+local queues = setmetatable({}, {
+  __newindex = function (self, name, queue)
+    return rawset(self, make_queue_key(name), queue)
+  end,
+  __index = function (self, name)
+    return rawget(self, make_queue_key(name))
+  end
+})
+
+
+function Queue.exists(name)
+  return queues[name] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_coalescing_delay` and `max_batch_size`
+-- @return table: a Queue object.
+local function get_or_create_queue(queue_conf, handler, handler_conf)
+
+  local name = queue_conf.name
+
+  local queue = queues[name]
+  if queue then
+    queue:log_debug("queue exists")
+    -- We always use the latest configuration that we have seen for a queue and handler.
+    queue.handler_conf = handler_conf
```

Comment:
Should we flush the queue before updating the conf? Or a previously enqueued item will be handled with new conf.

Reply:
We want the configuration update to be in effect immediately and affect previously queued items as well as new ones.

Reply:
Or we could simply bind the conf to the entity.

Reply:
This does not seems to match our previous discussions, and it alters the behavior significantly from the current implementation. The risk is that the queue name must be chosen carefully under the assumption that config might change from when item was enqueued to when item gets flushed, and such decision should probably be considered more carefully at such a global level.

Reply:
We could do that, but it would incur a memory cost and make it impossible to get rid of items that are not making progress due to a misconfiguration.  Imagine that a customer has set a high `max_retry_time` and they need to change the log server address because the old address no longer works.  They would have no way to get rid of already queues items in that case, other than waiting for the `max_retry_time` to expire and the entries being deleted.

Reply:
> This does not seems to match our previous discussions, and it alters the behavior significantly from the current implementation.

@dndx Can you explain what you refer to by "This"?

Reply:
> It certainly deserves mentioning that because we don't have an explicit `configure` callback in plugins that would be invoked when a reconfiguration is done, updated plugin parameters would only be put into effect when a new item is queued.

rate-limiting advanced solves this problem by registering for plugin curd events. The `configure` handler idea would also benefit at least: 1. exit-transformer, 2. rate-limiting advanced

Reply:
Are plugin crud events also issued in dbless and hybrid mode?

Reply:
@hanshuebner in dbless RAL register for configure update